### PR TITLE
Make push work with push.default=matching repos

### LINKED ISSUE #4888: Fail to push git-annex branch
EDIT: The issue also affects more standard workflows, see comments below

I have a dataset with data from a zipped tarball in a datalad-archives special remote (context is in https://github.com/datalad-handbook/book/pull/569#discussion_r483675080).
The dataset feels fine overall, but I can't push the git-annex branch anymore:

```
datalad create some && cd some

[INFO   ] Creating a new annex repo at /tmp/some 
create(ok): /tmp/some (dataset)
(handbook) ╭─adina@muninn /tmp/some on master
╰─➤ datalad download-url --archive 'https://s3.amazonaws.com/fast-ai-imageclas/imagenette2-160.tgz'
[INFO   ] Downloading 'https://s3.amazonaws.com/fast-ai-imageclas/imagenette2-160.tgz' into '/tmp/some/' 
download_url(ok): /tmp/some/imagenette2-160.tgz (file)                                                             
add(ok): /tmp/some/imagenette2-160.tgz (file)
save(ok): /tmp/some (dataset)
[INFO   ] Adding content of the archive /tmp/some/imagenette2-160.tgz into annex AnnexRepo(/tmp/some) 
[INFO   ] Initiating special remote datalad-archives 
[INFO   ] Finished adding /tmp/some/imagenette2-160.tgz: Files processed: 13394, +annex: 13394 
action summary:
  add (ok: 1)
  download_url (ok: 1)
  save (ok: 1)
(handbook) ╭─adina@muninn /tmp/some on master
╰─➤ datalad create-sibling ../anotherplace

[INFO   ] Considering to create a target dataset /tmp/some at /tmp/anotherplace of localhost 
[INFO   ] Fetching updates for Dataset(/tmp/some) 
[INFO   ] Adjusting remote git configuration                                                                       
[INFO   ] Running post-update hooks in all created siblings 
(handbook) ╭─adina@muninn /tmp/some on master
╰─➤ datalad push --to local
copy(ok): /tmp/some/imagenette2-160/train/n01440764/ILSVRC2012_val_00000293.JPEG (file) [to local...]              
copy(ok): /tmp/some/imagenette2-160/train/n01440764/ILSVRC2012_val_00002138.JPEG (file) [to local...]              
copy(ok): /tmp/some/imagenette2-160/train/n01440764/ILSVRC2012_val_00003014.JPEG (file) [to local...]              
copy(ok): /tmp/some/imagenette2-160/train/n01440764/ILSVRC2012_val_00006697.JPEG (file) [to local...]              
copy(ok): /tmp/some/imagenette2-160/train/n01440764/ILSVRC2012_val_00007197.JPEG (file) [to local...]              
copy(ok): /tmp/some/imagenette2-160/train/n01440764/ILSVRC2012_val_00009346.JPEG (file) [to local...]              
copy(ok): /tmp/some/imagenette2-160/train/n01440764/ILSVRC2012_val_00009379.JPEG (file) [to local...]              
copy(ok): /tmp/some/imagenette2-160/train/n01440764/ILSVRC2012_val_00009396.JPEG (file) [to local...]              
copy(ok): /tmp/some/imagenette2-160/train/n01440764/ILSVRC2012_val_00010306.JPEG (file) [to local...]              
copy(ok): /tmp/some/imagenette2-160/train/n01440764/ILSVRC2012_val_00011233.JPEG (file) [to local...]              
Update availability for 'local':  75%|███████████████████████████▊         | 3.00/4.00 [00:00<00:00, 12.7k Steps/s]CommandError: 'git push --progress --porcelain local refs/heads/git-annex:refs/heads/git-annex git-annex:git-annex' failed with exitcode 1 under /tmp/some
error: dst ref refs/heads/git-annex receives from more than one src.
error: failed to push some refs to '/tmp/anotherplace'

```

Even though the final push takes quite a while, nothing gets transferred to the sibling - the sibling is a Git repository with no commits yet.

The same error also happens without the change in #4877, and also if I push to a sibling that is a dataset clone added with ``datalad siblings add`` instead of one created with ``datalad create -sibling``.

<details><summary>Same issue without the change from #4877</summary>

```
(handbook) ╭─adina@muninn /tmp
╰─➤ datalad create some && cd some
[INFO   ] Creating a new annex repo at /tmp/some 
create(ok): /tmp/some (dataset)
(handbook) ╭─adina@muninn /tmp/some on master
╰─➤ datalad download-url --archive 'https://s3.amazonaws.com/fast-ai-imageclas/imagenette2-160.tgz'

[INFO   ] Downloading 'https://s3.amazonaws.com/fast-ai-imageclas/imagenette2-160.tgz' into '/tmp/some/' 
download_url(ok): /tmp/some/imagenette2-160.tgz (file)                                                             
add(ok): /tmp/some/imagenette2-160.tgz (file)
save(ok): /tmp/some (dataset)
[INFO   ] Adding content of the archive /tmp/some/imagenette2-160.tgz into annex AnnexRepo(/tmp/some) 
[INFO   ] Initiating special remote datalad-archives 
[INFO   ] Finished adding /tmp/some/imagenette2-160.tgz: Files processed: 1, +annex: 1 
action summary:
  add (ok: 1)
  download_url (ok: 1)
  save (ok: 1)
(handbook) ╭─adina@muninn /tmp/some on master
╰─➤ datalad create-sibling ../anotherplace
[INFO   ] Considering to create a target dataset /tmp/some at /tmp/anotherplace of localhost 
[INFO   ] Fetching updates for Dataset(/tmp/some) 
[INFO   ] Adjusting remote git configuration                                                                       
[INFO   ] Running post-update hooks in all created siblings 
create_sibling(ok): /tmp/some (dataset)
(handbook) ╭─adina@muninn /tmp/some on master
╰─➤ datalad push --to local
copy(ok): /tmp/some/MD5E-s98948031--0edfc972b5c9817ac36517c0057f3869.tar (file) [to local...]                      
Update availability for 'local':  75%|███████████████████████████▊         | 3.00/4.00 [00:00<00:00, 16.7k Steps/s]CommandError: 'git push --progress --porcelain local refs/heads/git-annex:refs/heads/git-annex git-annex:git-annex' failed with exitcode 1 under /tmp/some
error: dst ref refs/heads/git-annex receives from more than one src.
error: failed to push some refs to '/tmp/anotherplace'
```
</details>

see maybe also https://git-annex.branchable.com/forum/refs__47__heads__47__synced__47__git-annex_receives_from_more_than_one_src/

### PR DESCRIPTION
**Pull request by mih**
Story is in the commit messages. Big change is the push of the active branch.

Fixes #4888

### CODE REVIEW

File: datalad/core/distributed/push.py [hunk 1]
```diff
@@ -523,20 +533,17 @@ def _push(dspath, content, target, data, force, jobs, res_kwargs, pbars,
...
-        )
+
+    # make sure that we always push the active branch (the context for the
+    # potential path arguments) and the annex branch -- because we claim
+    # to know better than any git config
+    must_have_branches = [active_branch] if active_branch else []
+    if is_annex_repo:
+        must_have_branches.append('git-annex')
+    for branch in must_have_branches:
+        # active_branch could be None
```
Comment (kyleam) [inline datalad/core/distributed/push.py:544]: Not true with the current upstream code :)

**Review by kyleam — APPROVED** [review]
Okay, so when we repurposed the optimization in 211fead30 as a fix for gh-4780, we traded the gh-4780 issue/limitation for this bug.  (That commit was introduced in 0.13.2.)

The fix looks good to me.  As you note, the always-push-current is a change in behavior, but given how young `push` is and how unlikely it seems that someone would rely on the previous behavior in general, I think it's safe.

I pushed a fix for the new test on crippled fs.  It fixes the run for me locally under `tools/eval_under_testloopfs`.  I'm hoping it also fixes the appveyor failure.

**Comment by kyleam** [comment]
> I pushed a fix for the new test on crippled fs.

The build still had a push test failure, but I think it's one we've been seeing.  The new `test_push_matching` now passes.

https://github.com/datalad/datalad/pull/4896/checks?check_run_id=1081754243

File: datalad/core/distributed/push.py (hunk 1 above)
Comment (mih) [reply datalad/core/distributed/push.py:544]: Arrgh, true. Too many iterations if moving conditionals around. Will fix that up.

**Comment by mih** [comment]
Thx @kyleam 

I believe the singularity failure on the -container tests is unrelated to this PR. Will merge.

**Pull request merged**